File: indexer/state.py
from __future__ import annotations

import json
import time
from pathlib import Path
from typing import Any

from sqlalchemy import create_engine, select
from sqlalchemy.dialects.sqlite import insert as sqlite_insert
from sqlalchemy.orm import Session

from indexer.models import Base, RepoCache


class StateStore:
    def __init__(self, db_path: str) -> None:
        self.db_path = db_path
        self._engine = self._connect(db_path)
        self._ensure_schema()

    def _connect(self, db_path: str):
        db = Path(db_path)
        db.parent.mkdir(parents=True, exist_ok=True)
        return create_engine(f"sqlite:///{db}", future=True)

    def _ensure_schema(self) -> None:
        Base.metadata.create_all(self._engine)

    def should_refresh(
        self, repo: str, discovered_updated_at: str, stale_after_days: int
    ) -> bool:
        existing = self.read_repo(repo)
        if not existing:
            return True

        if existing["parse_error"]:
            return True

        if discovered_updated_at and existing["updated_at"] != discovered_updated_at:
            return True

        stale_seconds = max(1, stale_after_days) * 24 * 60 * 60
        return int(time.time()) - int(existing["scanned_at"]) >= stale_seconds

    def read_repo(self, repo: str) -> dict[str, Any] | None:
        with Session(self._engine) as session:
            row = session.get(RepoCache, repo)
            if row is None:
                return None

        payload: dict[str, Any] | None = None
        try:
            loaded = json.loads(row.payload_json)
            if isinstance(loaded, dict):
                payload = loaded
        except Exception:  # noqa: BLE001
            payload = None

        return {
            "repo": row.repo,
            "updated_at": row.updated_at,
            "scanned_at": row.scanned_at,
            "payload": payload,
            "parse_error": row.parse_error,
        }

    def list_payloads(self) -> list[dict[str, Any]]:
        with Session(self._engine) as session:
            rows = session.scalars(
                select(RepoCache).where(RepoCache.parse_error.is_(None))
            ).all()

        payloads: list[dict[str, Any]] = []
        for row in rows:
            try:
                loaded = json.loads(row.payload_json)
            except Exception:  # noqa: BLE001
                continue
            if isinstance(loaded, dict):
                payloads.append(loaded)
        return payloads

    def upsert_repo(
        self,
        repo: str,
        updated_at: str,
        payload: dict[str, Any],
        parse_error: str | None = None,
    ) -> None:
        now = int(time.time())
        stmt = sqlite_insert(RepoCache).values(
            repo=repo,
            updated_at=updated_at,
            scanned_at=now,
            payload_json=json.dumps(payload, ensure_ascii=True),
            parse_error=parse_error,
        )
        stmt = stmt.on_conflict_do_update(
            index_elements=[RepoCache.repo],
            set_={
                "updated_at": updated_at,
                "scanned_at": now,
                "payload_json": json.dumps(payload, ensure_ascii=True),
                "parse_error": parse_error,
            },
        )
        with Session(self._engine) as session:
            session.execute(stmt)
            session.commit()

    def close(self) -> None:
        self._engine.dispose()

File: src/state.ts
import { mkdirSync } from "fs";
import { dirname } from "path";
import { drizzle } from "drizzle-orm/better-sqlite3";
import Database from "better-sqlite3";
import { sqliteTable, text, integer } from "drizzle-orm/sqlite-core";
import { eq, isNull } from "drizzle-orm";
import type { ThemeEntry } from "./types.js";

const repoCache = sqliteTable("repo_cache", {
  repo: text("repo").primaryKey(),
  updatedAt: text("updated_at").notNull(),
  scannedAt: integer("scanned_at").notNull(),
  payloadJson: text("payload_json").notNull(),
  parseError: text("parse_error"),
});

export interface RepoCacheEntry {
  repo: string;
  updated_at: string;
  scanned_at: number;
  payload: ThemeEntry | Record<string, unknown> | null;
  parse_error: string | null;
}

export class StateStore {
  private db: ReturnType<typeof drizzle>;
  private sqlite: Database.Database;

  constructor(dbPath: string) {
    mkdirSync(dirname(dbPath), { recursive: true });
    this.sqlite = new Database(dbPath);
    this.db = drizzle(this.sqlite);
    this.init();
  }

  get raw(): Database.Database {
    return this.sqlite;
  }

  init(): void {
    this.sqlite.exec(`
      CREATE TABLE IF NOT EXISTS repo_cache (
        repo TEXT PRIMARY KEY,
        updated_at TEXT NOT NULL,
        scanned_at INTEGER NOT NULL,
        payload_json TEXT NOT NULL,
        parse_error TEXT
      )
    `);
  }

  close(): void {
    this.sqlite.close();
  }

  readRepo(repo: string): RepoCacheEntry | null {
    const rows = this.db
      .select()
      .from(repoCache)
      .where(eq(repoCache.repo, repo))
      .all();

    const row = rows[0];
    if (!row) {
      return null;
    }

    let payload: ThemeEntry | Record<string, unknown> | null = null;
    try {
      const loaded = JSON.parse(row.payloadJson);
      if (typeof loaded === "object" && loaded !== null) {
        payload = loaded;
      }
    } catch {
      payload = null;
    }

    return {
      repo: row.repo,
      updated_at: row.updatedAt,
      scanned_at: row.scannedAt,
      payload,
      parse_error: row.parseError,
    };
  }

  upsertRepo(
    repo: string,
    updatedAt: string,
    payload: ThemeEntry | Record<string, unknown>,
    parseError: string | null = null
  ): void {
    const now = Math.floor(Date.now() / 1000);
    const payloadJson = JSON.stringify(payload);

    this.db
      .insert(repoCache)
      .values({
        repo,
        updatedAt,
        scannedAt: now,
        payloadJson,
        parseError,
      })
      .onConflictDoUpdate({
        target: repoCache.repo,
        set: {
          updatedAt,
          scannedAt: now,
          payloadJson,
          parseError,
        },
      })
      .run();
  }

  shouldRefresh(
    repo: string,
    discoveredUpdatedAt: string,
    staleAfterDays: number
  ): boolean {
    const existing = this.readRepo(repo);
    if (!existing) {
      return true;
    }

    if (existing.parse_error) {
      return true;
    }

    if (discoveredUpdatedAt && existing.updated_at !== discoveredUpdatedAt) {
      return true;
    }

    const staleSeconds = Math.max(1, staleAfterDays) * 24 * 60 * 60;
    const now = Math.floor(Date.now() / 1000);
    return now - existing.scanned_at >= staleSeconds;
  }

  listPayloads(): ThemeEntry[] {
    const rows = this.db
      .select({ payloadJson: repoCache.payloadJson })
      .from(repoCache)
      .where(isNull(repoCache.parseError))
      .all();

    const payloads: ThemeEntry[] = [];
    for (const row of rows) {
      try {
        const loaded = JSON.parse(row.payloadJson);
        if (typeof loaded === "object" && loaded !== null) {
          payloads.push(loaded as ThemeEntry);
        }
      } catch {
        continue;
      }
    }
    return payloads;
  }
}
